src/chart/helper/multipleGraphEdgeHelper.js

```diff
@@ -0,0 +1,221 @@
+
+    linkMap[key] = linkMap[key] || [];
+    linkMap[key].push(index);
+};
+
+/**
+ * get linkMap with key
+ * @param link
+ * @param seriesModel
+ * @param index
+ */
+var getLinkFromMap = function (link, seriesModel) {
+    var key = getKeyOfLinks(link, seriesModel);
+    return linkMap[key];
+};
+
+/**
+ * calculate all cases total length
+ * @param link
+ * @param seriesModel
+ * @returns {number}
+ */
+var getTotalLengthBetweenNodes = function (link, seriesModel) {
+    var len = getLinkMapLengthWithKey([seriesModel.uid, link.source, link.target].join('>'));
+    var lenV = getLinkMapLengthWithKey([seriesModel.uid, link.target, link.source].join('>'));
+
+    return len + lenV;
+};
+
+/**
+ *
+ * @param key
+ */
+var getLinkMapLengthWithKey = function (key) {
+    return linkMap[key] ? linkMap[key].length : 0;
+};
+
+/**
+ * Count the number of edges between the same two points, used to obtain the curvature table and the parity of the edge
+ * @param edges
+ * @param seriesModel
+ * @param graph
+ */
+export function calculateMutilEdges(edges, seriesModel, graph) {
+    if (!getAutoCurvenessParams(seriesModel)) {
+        return;
+    }
+    // Hang on this object 4 dispose
+    curvenessList = graph.curvenessList || [];
+    linkMap = graph.linkMap = graph.linkMap || {};
+
+
+    for (var i = 0; i < edges.length; i++) {
+        var link = edges[i];
+        var source = link.source;
+        var target = link.target;
+        if (!source || !target) {
+            continue;
+        }
+        setLinkToMap(link, seriesModel, i);
+    }
+
+    // calc the array of curvenessList
+    createCurveness(seriesModel);
+}
+
+/**
+ * Set curvature for link
+ * @param link
+ * @param seriesModel
+ * @param index
+ */
+export function setCurvenessForLink(link, seriesModel, index) {
+    if (!getAutoCurvenessParams(seriesModel)) {
+        return;
+    }
+
+    var linkArray = getLinkFromMap(link, seriesModel);
+    if (!linkArray) {
+        return;
+    }
+
+    var linkIndex = linkArray.findIndex(function (it) {
+        return it === index;
+    });
+    var totalLen = getTotalLengthBetweenNodes(link, seriesModel);
+    // if totalLen bigger than curvenessList, recreate curvenessList
+    if (totalLen > curvenessList.length) {
+        createCurveness(seriesModel, totalLen);
+    }
+
+    link.lineStyle = link.lineStyle || {};
+    // if is opposite link, must set curvenss to opposite number
+    var curKey = getKeyOfLinks(link, seriesModel);
+    if (!linkArray.isForward) {
+        // the opposite link show outside
+        var oppositeKey = getOppositeKey(curKey);
+        var len = getLinkMapLengthWithKey(oppositeKey);
+        var layout = seriesModel.getModel('layout').option;
+        // Because the curvature algorithm of each layout is different, the reverse needs to be adapted here
+        if (layout === 'none' || layout === 'force') {
+            link.lineStyle.curveness = -1 * curvenessList[linkIndex + len + (totalLen % 2 ? 0 : 1)];
```

Comment:
Theoretically `link.lineStyle` should better not be modified internally, because user input (option) will be merged into it each time `setOption` called. If we modify it internally because of "autoCurveness" is on, we can not roll back it to the original state if "autoCurveness" is turned off. So keep it "immutable" is better than modify it directly.

But if we keep it immutable, where do we store the result calculated curveness? 
I think if we `setCurvenessForLink` just at the position where `edge.getModel().get('lineStyle.curveness')` is called, there will be no need to store the calculated curveness. For example, it could be like:
```js
var curveness = zrUtil.retrieve3(
    edge.getModel().get('lineStyle.curveness');
    // rename `setCurvenessForLink` to `getCurvenessForLink` and return the result curveness.
    getCurvenessForLink(link, seriesModel, index),
    0
};
```

The place of `edge.getModel().get('lineStyle.curveness')` called are:
https://github.com/apache/incubator-echarts/blob/4.8.0/src/chart/graph/circularLayoutHelper.js#L77
https://github.com/apache/incubator-echarts/blob/4.8.0/src/chart/graph/forceLayout.js#L91
https://github.com/apache/incubator-echarts/blob/4.8.0/src/chart/graph/simpleLayoutHelper.js#L39